pkg/oci/repository.go

```diff
@@ -113,215 +109,201 @@ func NewRepository(repositoryName string, opts ...Option) (*Repository, error) {
 	return &Repository{repo, opt}, nil
 }
 
-// FetchBlobFrom fetches a blob from the given remote. It accepts a string
-// and a boolean to indicate whether the blob should be verified. In order to verify the blob,
-// the compressed blob is read and the digest is verified against the digest in the string.
-// The uncompressed blob is returned as a io.ReadCloser.
-// The fetched blob is cached if specified in the repository. Even if caching fails, the blob is still returned.
-func (r *Repository) FetchBlobFrom(s string, verify, cache bool, opts ...Option) (io.ReadCloser, error) {
-	opt, err := makeOptions(opts...)
+// PushResource takes a resource, reference path and identity information and caches a resource in the internal OCI registry.
+func (c *OCIClient) PushResource(ctx context.Context, opts ResourceOptions) error {
+	reader, err := c.ocmClient.GetResource(ctx, opts.ComponentVersion, opts.Resource, opts.ReferencePath)
 	if err != nil {
-		return nil, fmt.Errorf("failed to make options: %w", err)
+		return fmt.Errorf("failed to get resource: %w", err)
 	}
-	l, errf := r.fetchBlobFrom(s, verify, cache, opt.remoteOpts...)
-	if l != nil {
-		uncompressed, err := l.Uncompressed()
-		if err != nil {
-			return nil, kerrors.NewAggregate([]error{errf, err})
-		}
-		return uncompressed, errf
+	defer reader.Close()
+
+	if _, err := c.cacheResource(ctx, opts.ComponentVersion, opts.Resource, opts.Owner, opts.SnapshotName, reader); err != nil {
+		return fmt.Errorf("failed to cache resource '%s': %w", opts.Resource.GetName(), err)
 	}
-	return nil, errf
+
+	return nil
 }
 
-func (r *Repository) fetchBlobFrom(s string, verify, cache bool, opts ...remote.Option) (v1.Layer, error) {
-	ref, err := name.NewDigest(s)
-	if err != nil {
-		return nil, fmt.Errorf("%s: blob s must be of the form <name@digest>", ref)
+// FetchAndCacheResource fetches and then caches a resource. It will only fetch a resource if it doesn't already have a snapshot.
+func (c *OCIClient) FetchAndCacheResource(ctx context.Context, opts ResourceOptions) (io.ReadCloser, error) {
```

Comment:
So, there are several reasons why snapshot handling is done here.

## Caching

Snapshots are used for caching / note / find out if a cache entry exists. I think this could potentially be done with the Repository's name, but that doesn't scale nicely.

If we didn't do that here, the client would have to do this check first, which makes no sense from the client's perspective. Why would I check if a snapshot exists or not? Ideally, I don't even care about snapshots at all.

## Transparent Snapshot

The client doesn't care about snapshots unless it must. If we push the snapshot handling onto the client, we risk further complicating client code and duplicating logic which retrieves a snapshot in case it exists.

Ideally, the client should be handling/caring about snapshots at all. Unless they explicitly want to do so.

## Client Fetching Snapshots

The client is seldom aware that it needs a snapshot in this scenario. You either define it or not. There is a bit of code in Localization/Configuration now which decides which source to use, but otherwise, calling FetchAndCache shouldn't have to have an extra thing to perform. It would mean that FetchAndCache would have to return a bunch of things for the client to construct a snapshot, and the client would have to know about the naming scheme to name the snapshots the way that the cache finds them on any further calls.

Potentially, this would also mean that other clients, like the `remote-controller`, after calling a convenient one-operation method, now would have to also create a snapshot for some reason they shouldn't care about/shouldn't be aware of in the first place.

## Conclusion

I understand what you are saying. It would be nice if the OCI code would be fully localised and wouldn't have to care about anything other than the registry. But in reality, I think that complicates a lot of the calling code that doesn't need to be complicated if the OCI registry does these things.

That said, I'll bow to consensus if you both think it would be better to push the snapshot handling onto the clients.